cmd/quadlet/main.go

```diff
@@ -0,0 +1,246 @@
+package main
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/containers/quadlet/pkg/quadlet"
+	"github.com/containers/quadlet/pkg/systemdparser"
+)
+
+var verboseFlag bool
+var versionFlag bool
+var isUser bool
+
+var noKmsg bool = false
+var kmsgFile *os.File = nil
+
+func logToKmsg(s string) bool {
+	if noKmsg {
+		return false
+	}
+
+	if kmsgFile == nil {
+		f, err := os.OpenFile("/dev/kmsg", os.O_WRONLY, 0644)
+		if err != nil {
+			noKmsg = true
+			return false
+		}
+		kmsgFile = f
+	}
+
+	if _, err := kmsgFile.Write([]byte(s)); err != nil {
+		kmsgFile.Close()
+		kmsgFile = nil
+		return false
+	}
+
+	return true
+}
+
+func Log(format string, a ...any) {
+	s := fmt.Sprintf(format, a...)
+	line := fmt.Sprintf("quadlet-generator[%d]: %s", os.Getpid(), s)
+
+	if !logToKmsg(line) {
+		// If we can't log, print to stderr
+		fmt.Fprintf(os.Stderr, "%s\n", line)
+		os.Stderr.Sync()
+	}
+}
+
+var debugEnabled bool = false
+
+func enableDebug() {
+	debugEnabled = true
+}
+
+func Debug(format string, a ...any) {
+	if debugEnabled {
+		Log(format, a...)
+	}
+}
+
+func getUnitDirs(user bool) []string {
+
+	unitDirsEnv := os.Getenv("QUADLET_UNIT_DIRS")
+	if len(unitDirsEnv) > 0 {
+		return strings.Split(unitDirsEnv, ":")
+	}
+
+	dirs := make([]string, 0)
+	if user {
+		if configDir, err := os.UserConfigDir(); err == nil {
+			dirs = append(dirs, path.Join(configDir, "containers/systemd"))
+		}
+	} else {
+		dirs = append(dirs, quadlet.UnitDirAdmin)
+		dirs = append(dirs, quadlet.UnitDirDistro)
+	}
+	return dirs
+}
+
+func loadUnitsFromDir(sourcePath string, units map[string]*systemdparser.UnitFile) {
+	files, err := os.ReadDir(sourcePath)
+	if err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			Log("Can't read \"%s\": %s", sourcePath, err)
+		}
+		return
+	}
+
+	for _, file := range files {
+		name := file.Name()
+		if units[name] == nil &&
```

Comment:
I think Quadlet should log when `units[name] != nil` (i.e., a unit-name clash across multiple directories).  It may be a simple oversight of the user who may be tricked into believing "it just works".  Debugging that scenario can be time intensive as there will be a service running with the expected name but with an unexpected workload.

Reply:
This is how all systemd overriding works though. If you have a /usr/ version of the .container file you can override that as sysadmin by adding one in /etc.